### PR TITLE
Add support for firmware versions

### PR DESCRIPTION
**Pull request by jexposit**
Patches from Peter (https://github.com/linuxwacom/libwacom/pull/620) plus changes from my side.

Requires https://lore.kernel.org/linux-input/20240322100210.107152-1-jose.exposito89@gmail.com/T/ or https://github.com/JoseExposito/digimend-kernel-drivers/tree/sysfs-fw-name.

Fixes https://github.com/linuxwacom/libwacom/issues/610

cc @whot

### CODE REVIEW

File: data/wacom.example [hunk 1]
```diff
@@ -74,14 +74,25 @@ ModelName=
...
+#
+# Multiple matches are joined by a semicolon
+DeviceMatch=usb|056a|00bc;bluetooth|056a|00bc;
+# If the vendor/product ID is not sufficient to distinguish the name
+# (common for HUION, GAOMON, UCLogic, XP-Pen), use the device name
+# as fourth element in the match string.
+# DeviceMatch=usb|1234|abcd:SomeDevice Name;
+#
+# Optionally after the device name (which may be empty) the
+# the remainer of that substring is the firmware version to match against.
```
Comment (whot) [inline data/wacom.example:86]:
```suggestion
# the remainder of that substring is the firmware version to match against.
```

File: libwacom/libwacom.c [hunk 2]
```diff
@@ -212,11 +212,43 @@ client_query_by_subsystem_and_device_file (GUdevClient *client,
...
+		return NULL;
+
+	/* The UCLogic kernel driver returns firmware names with form
+	 * <vendor>_<model>_<version>. Remove the version from `fw` to avoid
+	 * mismatches on firmware updates. */
+	regex = g_regex_new ("(.*_.*)_.*$",
+				G_REGEX_DEFAULT,
+				G_REGEX_MATCH_DEFAULT,
+				NULL);
+	g_regex_match (regex, fw, 0, &match_info);
```
Comment (whot) [inline libwacom/libwacom.c:232]: Note to self: this may become an issue if we ever get a fw exported that has a different underscorify pattern.

**Review by whot — APPROVED** [review]
One typo found, would be great if you could squash that in. Otherwise this LGTM, we can merge this once upstream accepts your kernel patch. Can you ping me when that happens? Thanks

**Comment by jexposit** [comment]
@whot the [kernel patch has been merged](https://lore.kernel.org/linux-input/20240322100210.107152-1-jose.exposito89@gmail.com/T/#e66606679b7c120427b34cecbb5fe9ff4d921c1d5).

In the end, instead of creating a custom sysfs property, we ended up using `hid->uniq`. Since this a generic attribute all devices has, I wonder if it makes sense to refactor a little bit the series to use `uniq` instead of `fw` as variable names and update the documentation.

Something else I have in my to-do list is displaying that information with `libwacom-list-local-devices` so user can actually discover it.

**Comment by whot** [comment]
> I wonder if it makes sense to refactor a little bit the series to use uniq instead of fw as variable names and update the documentation.

yes, definitely. This also leaves room for a future firmware attribute if uniq is no longer sufficient. 

> Something else I have in my to-do list is displaying that information with libwacom-list-local-devices so user can actually discover it.

I think `--format=datafile` should work automatically, the yaml format will need an extra entry though.

**Comment by jexposit** [comment]
> > I wonder if it makes sense to refactor a little bit the series to use uniq instead of fw as variable names and update the documentation.
> 
> yes, definitely. This also leaves room for a future firmware attribute if uniq is no longer sufficient.

Refactored the last patch to use uniq instead of fw

> > Something else I have in my to-do list is displaying that information with libwacom-list-local-devices so user can actually discover it.
> 
> I think `--format=datafile` should work automatically, the yaml format will need an extra entry though.

Also added an extra entry to `tools/list-devices.c`

File: libwacom/libwacom.c [hunk 3]
```diff
@@ -817,6 +876,7 @@ static void print_integrated_flags_for_device (int fd, const WacomDevice *device
 static void print_match(int fd, const WacomMatch *match)
 {
 	const char  *name       = libwacom_match_get_name(match);
+	const char  *uniq         = libwacom_match_get_uniq(match);
```
Comment (whot) [inline libwacom/libwacom.c:879]:
```suggestion
	const char  *uniq        = libwacom_match_get_uniq(match);
```

**Review by whot — APPROVED** [review]
one indentation error, rest looks good, thanks!

File: libwacom/libwacom.sym [hunk 4]
```diff
@@ -72,3 +72,7 @@ local:
 LIBWACOM_2.9 {
     libwacom_get_num_keys;
 } LIBWACOM_2.0;
+
+LIBWACOM_2.11 {
```
Comment (whot) [inline libwacom/libwacom.sym:76]:
```suggestion
LIBWACOM_2.12 {
```

(because I just released 2.11, sorry)

**Comment by jexposit** [comment]
Resolved and rebased. Thanks for the review!

**Comment by whot** [comment]
`../libwacom/libwacom.c:234:33: error: ‘G_REGEX_DEFAULT’ undeclared (first use in this function); did you mean ‘G_REGEX_DOTALL’?`

Introduced in 2.74 and that's too new for Ubuntu 22.04. Looking at glib commit `2.73.1-26-g879b9cd669f0` we should be fine with an
```
#ifndef G_REGEX_DEFAULT
#define G_REGEX_DEFAULT 0
#endif
```

**Comment by jexposit** [comment]
Ops, missed the CI emails. I replaced `compile_options` and `match_options` with 0 as in the docs:
https://docs.gtk.org/glib/ctor.Regex.new.html

**Comment by jexposit** [comment]
Is the FreeBSD failure expected?

**Comment by whot** [comment]
> Is the FreeBSD failure expected?

yeah, it's very unreliable and fails often (all the time?) but I'm ETIME on debugging that. cc @jbeich

**Comment by whot** [comment]
I think we're good here, let's merge this! thanks for all the effort

**Comment by jbeich** [comment]
> it's very unreliable

Try applying https://github.com/labwc/labwc/commit/00ec29c834b7. I suspect `v0` isn't maintained anymore or emulating on macOS is still unstable.

**Comment by whot** [comment]
@jbeich thanks, that was it, fixed in #662

**Pull request merged**